### PR TITLE
#161966863 Install and Setup Redux

### PR DESCRIPTION
**Pull request by ThaDeveloper**
[Build Status](https://travis-ci.org/andela/ah-infinity-stones-frontend)
#### What does this PR do?
- Setup Redux

#### Description of Task to be completed?
- Install redux, react-redux and redux-saga
- Create appropriate redux folder structure
- Create sample component and apply middleware sagas.

#### How should this be manually tested?
1. `$ git clone -b ch-setup-redux-161966863 https://github.com/andela/ah-infinity-stones-frontend.git`  and `$ cd ah-infinity-stones-frontend`
2. Install the dependencies - `$ npm install`
3. Start the app by running `$ npm start` - This will run the app in the development mode.  Open http://localhost:3000 to view it in the browser.
The page will reload if you make edits. 	You will also see any lint errors in the console.
4. Test the app by running `$ npm test` - This will Launch the test runner in the interactive watch mode.

#### What are the relevant pivotal tracker stories?
[#161966863](https://www.pivotaltracker.com/story/show/161966863)

#### Screenshots
screenshot 2019-01-09 at 15 14 01

### CODE REVIEW

**Comment by rkemmy** [comment]
@ThaDeveloper please include a screenshot of package.json so that we can see the dependencies installed. Also the reducers/index.js should have a trailing line. ESLint has flagged that. Great work

**Comment by kamyaD** [comment]
@ThaDeveloper this looks great.

**Comment by PromasterGuru** [comment]
Good job @ThaDeveloper, it would also be advisable to add the container folder in the structure.

**Comment by ThaDeveloper** [comment]
> @ThaDeveloper please include a screenshot of package.json so that we can see the dependencies installed. Also the reducers/index.js should have a trailing line. ESLint has flagged that. Great work

Eslint has been solved, for the dependencies, I bet it's better to check the files changed for a cleaner precise PR with a limited amount of screenshots.

**Review by leonardnjura — COMMENTED** [review]
Nice initial setup @ThaDeveloper  We need this for a real world or large app

**Comment by ThaDeveloper** [comment]
> Good job @ThaDeveloper, it would also be advisable to add the container folder in the structure.

I don't think we need a container folder, based on what I know a container is simply a smart component, it's not independent of the component. @johnmutuma5  can add more light to this.

**Comment by codeplus254** [comment]
Great work! Everything is running without errors

**Review by codeplus254 — APPROVED** [review]
Looks great. Just ran it without errors

**Review by PromasterGuru — APPROVED** [review]

**Review by johnmutuma5 — CHANGES_REQUESTED** [review]
Justin please set an appropriate badge on the PR and check out [this comment](https://www.pivotaltracker.com/story/show/161966863/comments/198193524):

**Comment by ThaDeveloper** [comment]
@johnmutuma5 This has been done, thanks

**Comment by kamyaD** [comment]
@ThaDeveloper Good effort, however, please check  on Travis, it seems it's not passing.

**Comment by ThaDeveloper** [comment]
> @ThaDeveloper Good effort, however, please check on Travis, it seems it's not passing.

This has been fixed, the fail was being caused by a failing test caused by the `__snapshots__`  which was supposed to be git ignored.

**Review by kamyaD — APPROVED** [review]
Thanks for fixing travis

File: package.json [hunk 1]
```diff
@@ -5,7 +5,11 @@
   "dependencies": {
     "react": "^16.7.0",
     "react-dom": "^16.7.0",
-    "react-scripts": "2.1.3"
+    "react-redux": "^6.0.0",
+    "react-scripts": "2.1.3",
+    "redux": "^4.0.1",
+    "redux-devtools-extension": "^2.13.7",
```
Comment (johnmutuma5) [inline package.json]: @ThaDeveloper, this seems like it should be a devDependency not for production environment. 🤔

File: src/components/Articles/Articles.jsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    }
+  
+}
+const mapStateToProps = state => ({ increment: state.articleReducer.count });
+
+const mapDispatchToProps = dispatch =>
+  bindActionCreators({ requestIncrement }, dispatch);
+
+//make our component smart
+export default connect(mapStateToProps, mapDispatchToProps)(Articles);
```
Comment (johnmutuma5) [inline src/components/Articles/Articles.jsx]: @ThaDeveloper, nice work on this redux store connection. We can improve it by just passing the `actionCreators` object in the place of `mapDispatchToProps` and redux is going to call the `bindActionCreators` function for us internally.

```js
    // ...
    const actionCreators = { requestIncrement };
    export default connect(mapStateToProps, actionCreators)(Articles); // 🙂
```

File: src/redux/store.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { applyMiddleware, createStore, compose } from 'redux';
```
Comment (johnmutuma5) [inline src/redux/store.js:1]: @ThaDeveloper, great work on this file. 👏

File: src/redux/sagas/rootSaga.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { put, takeLatest } from "redux-saga/effects";
```
Comment (johnmutuma5) [inline src/redux/sagas/rootSaga.js]: This file is not really the rootSaga, it is probably the `incrementSaga.js` as it contain the `incrementSaga` and the `incrementRequestWatcher`. The watcher to listen for the `REQUEST_INCREMENT` action and the `incrementSaga` to handle the action. 🤔 

The` rootSaga` on the other hand should be created separately(*probably like how you put the `rootReducer` in the `index.js` of its package*). The work of the `rootSaga` is to `yield all` the `action watchers`, like `incrementRequestWatcher`, in the application and it is what runs in `middleware.run(rootSaga)` as you did in `store.js`.

File: src/redux/sagas/rootSaga.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+import { receiveIncrement } from "../actions/articleActions";
+import { REQUEST_INCREMENT } from "../actions/actionTypes";
+
+// worker Saga: will be fired once user request actions
+function* increment(action) {
```
Comment (johnmutuma5) [inline src/redux/sagas/rootSaga.js]: This is the `saga`. We should probably name it `incrementSaga` for more clarity. It acts as the slave that receives instructions from any `watcher` that has registered it.

File: src/redux/sagas/rootSaga.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  try {
+    // do an api call here
+    yield put(receiveIncrement());
+  } catch (e) {
+    console.log(e);
+  }
+}
+
+// once the REQUEST_INCREMENT action type is called, increment method is run
+export default function* saga() {
```
Comment (johnmutuma5) [inline src/redux/sagas/rootSaga.js]: This is the `watcher` which should be yielded amongst other `watchers` in the `rootSaga`, that we should create as pointed out earlier. It's work is to listen for `actions`; once it catches an `action`, it instructs the slave `saga` that it has registered in `takeLatest` below, in this case the `incrementSaga`, handing it the action for processing(*as you saw, the `incrementSaga` takes `action` as argument*).

I hope the flow is getting clearer.

File: src/redux/store.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import createSagaMiddleware from 'redux-saga';
+import allReducers from './reducers';
+import saga from "./sagas/rootSaga";
```
Comment (johnmutuma5) [inline src/redux/store.js]: Remember to update this import once you decide where to put the `rootSaga` (*or `allSagas` to be consistent with how you have named `allReducers`*) 🙂

**Review by johnmutuma5 — CHANGES_REQUESTED** [review]
Great work on the set up. Wonderful progress in such a short span of time @ThaDeveloper, I have left some reviews.

File: src/redux/sagas/rootSaga.js (hunk 4 above)
Comment (ThaDeveloper) [reply src/redux/sagas/rootSaga.js]: Sure,totally makes sense

File: src/redux/sagas/rootSaga.js (hunk 5 above)
Comment (ThaDeveloper) [reply src/redux/sagas/rootSaga.js]: Cool

**Comment by ThaDeveloper** [comment]
@johnmutuma5 All the changes have been made, kindly have a look

File: src/redux/store.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import { applyMiddleware, createStore, compose } from 'redux';
+import createSagaMiddleware from 'redux-saga';
+import allReducers from './reducers';
+import rootSaga from "./sagas/rootSaga";
```
Comment (johnmutuma5) [inline src/redux/store.js]: @ThaDeveloper, for consistency, I suggest we put the `rootSaga` in the `index.js` of the sagas like how we have done with the `allReducers`; unless you can foresee a problem with that.

**Review by johnmutuma5 — CHANGES_REQUESTED** [review]
@ThaDeveloper, this is incredible. Very good setup. Please respond to one more comment below and we'll be done with this one. Thanks.

File: src/redux/store.js (hunk 8 above)
Comment (ThaDeveloper) [reply src/redux/store.js]: okay, no problem

**Review by johnmutuma5 — APPROVED** [review]
Great work. @ThaDeveloper

**Pull request merged**